src/tools/ceph_dedup_tool.cc

```diff
@@ -551,45 +552,110 @@ class SampleDedupWorkerThread : public Thread
   public:
     using dup_count_t = ssize_t;
 
-    bool find(string& fp) {
+    auto find(string& fp) {
+      auto iter = under_threshold_fp_map.find(fp);
+      if (iter == under_threshold_fp_map.end()) {
+	return over_threshold_fp_map.find(fp);
+      } 
+      return iter;
+    }
+
+    auto is_end() {
+      return over_threshold_fp_map.end();
+    }
+
+    bool contain(string& fp) {
       std::shared_lock lock(fingerprint_lock);
-      auto found_item = fp_map.find(fp);
-      return found_item != fp_map.end();
+      auto found_item = find(fp);
+      return found_item != is_end();
+    }
+
+    size_t fp_map_size() const {
+      return under_threshold_fp_map.size() + over_threshold_fp_map.size();
+    }
+
+    bool fp_map_empty() const {
+      return (under_threshold_fp_map.size() == 0 && over_threshold_fp_map.size() == 0);
+    }
+
+    size_t entry_length() const {
+      if (under_threshold_fp_map.empty()) {
+	if (over_threshold_fp_map.empty()) {
+	  return 0;
+	}
+	return over_threshold_fp_map.begin()->first.length();
+      }
+      return under_threshold_fp_map.begin()->first.length();
     }
 
     // return true if the chunk is duplicate
     bool add(chunk_t& chunk) {
+      enum class fp_store_state_t {
+	NORMAL,
+	FULL
+      } state = fp_store_state_t::NORMAL;
+
       std::unique_lock lock(fingerprint_lock);
-      auto found_iter = fp_map.find(chunk.fingerprint);
+      auto found_iter = find(chunk.fingerprint);
       ssize_t cur_reference = 1;
-      if (found_iter == fp_map.end()) {
-        fp_map.insert({chunk.fingerprint, 1});
+      if (found_iter == is_end()) {
+	if (state != fp_store_state_t::FULL) {
+	  under_threshold_fp_map.insert({chunk.fingerprint, 1});
+	}
+	if (is_fpmap_full()) {
+	  auto len_before = fp_map_size();
+	  try_shrink();
```

Comment:
This appears to be the only caller of try_shrink(), just replace it with under_threshold_map.clear().  You don't need len_before and len_after, it's just the size of the over_threshold_fp_map.